### PR TITLE
Added the function construct_pdp_query to build cypher queries that c…

### PR DESCRIPTION
**Pull request by ben-heil**
Construct_pdp_query acts like construct_dwpc_query, except that it returns the path degree product instead. This function also takes in the dwpc so that the path degree products can be represented as a fraction of the total dwpc. If the dwpc is not provided, a subquery will be added to calculate it.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -22,6 +22,10 @@
 tests_require = [
     'numpy',
     'scipy',
+    'neo4j',
+    'py2neo == 3',
+    'pandas',
+    'tqdm'
```
Comment (zietzm) [inline setup.py]: Is `tqdm` still a requirement? I can't seem to find an import anywhere

File: hetio/neo4j.py [hunk 2]
```diff
@@ -281,6 +281,171 @@ def construct_dwpc_query(metarels, property='name', join_hint='midpoint', index_
 
     return query
 
+def construct_pdp_query(metarels, dwpc=None, property='name', join_hint='midpoint', index_hint=False, unique_nodes=True):
+    """
+    Create a cypher query for computing the path degree product for a type of path.
```
Comment (zietzm) [inline hetio/neo4j.py]: Really minor, but I think `cypher` should be capitalized (`Cypher`)

File: hetio/neo4j.py [hunk 3]
```diff
@@ -281,6 +281,171 @@ def construct_dwpc_query(metarels, property='name', join_hint='midpoint', index_
...
+    join_hint : 'midpoint', bool, or int
+        whether to add a join hint to tell neo4j to traverse form both ends of
+        the path and join at a specific index. `'midpoint'` or `True` specifies
+        joining at the middle node in the path (rounded down if an even number
+        of nodes). `False` specifies not to add a join hint. An int specifies
+        the node to join on.
+    index_hint : bool
+        whether to add index hints which specifies the properties of the source
+        and target nodes to use for lookup. Enabling both `index_hint` and
+        `join_hint` can cause the query to fail.
```
Comment (zietzm) [inline hetio/neo4j.py]: Is an integer `join_hint` equivalent to an `index_hint`?

File: hetio/neo4j.py [hunk 4]
```diff
@@ -281,6 +281,171 @@ def construct_dwpc_query(metarels, property='name', join_hint='midpoint', index_
...
+            'i0': i,
+            'i1': i + 1,
+            'source_label': source_label,
+            'target_label': target_label,
+            'rel_type': rel_type,
+            'dir0': '<-' if direction == 'backward' else '-',
+            'dir1': '->' if direction == 'forward' else '-',
+        }
+        degree_strs.append(textwrap.dedent(
+            '''\
```
Comment (zietzm) [inline hetio/neo4j.py]: Really nice. Had to look up the trailing backslash because I didn't realize this would omit newlines.

File: test/neo4j_test.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    """
+    Test dwpc query construction and computation on the metapath from 
+    https://doi.org/10.1371/journal.pcbi.1004259.g002
+    """
+
+    directory = pathlib.Path(__file__).parent.absolute()
+    path = directory.joinpath('data/bupropion-CbGpPWpGaD-subgraph.json.gz')
+
+    graph = hetio.readwrite.read_graph(path)
+    assert graph is not None
```
Comment (zietzm) [inline test/neo4j_test.py]: Always good to err on the side of extra testing, but I'm curious, were you encountering instances in which `graph` was returned as `None`?

File: setup.py (hunk 1 above)
Comment (zietzm) [reply setup.py]: My mistake! I see it is in the `neo4j.py` file.

File: hetio/neo4j.py (hunk 2 above)
Comment (ben-heil) [reply hetio/neo4j.py]: You're right, I'll go ahead and update it

File: hetio/neo4j.py (hunk 3 above)
Comment (ben-heil) [reply hetio/neo4j.py]: My understanding is that index hints tell the Cypher query planner to start searching at the supplied index, while join hints tell the planner to start in multiple places and meet at the join index. That being said, this was part of the code that I pulled from construct_dwpc_query, so you'd have to ask Daniel to be sure.

File: hetio/neo4j.py (hunk 4 above)
Comment (ben-heil) [reply hetio/neo4j.py]: This was also courtesy of construct_dwpc_query, but yeah, backslashes in strings do cool things

File: test/neo4j_test.py (hunk 5 above)
Comment (ben-heil) [reply test/neo4j_test.py]: Currently when the path doesn't yield any results, read_graph throws a KeyError so the test fails before getting here. That being said, this assertion acts as a sanity check in case something causes read_graph to eat the exception

File: hetio/neo4j.py [hunk 6]
```diff
@@ -281,6 +281,171 @@ def construct_dwpc_query(metarels, property='name', join_hint='midpoint', index_
...
+            metapath_query = metapath_query,
+            using_query = using_query,
+            unique_nodes_query = unique_nodes_query,
+            degree_query = degree_query,
+            length=len(metarels),
+            property=property,
+            dwpc = dwpc)
+    # If the dwpc isn't provided, we'll have to calculate it before the PDP.
+    # Doing so roughly doubles the query execution time, as it effectively
+    # runs the query twice returning different degrees of aggregation.
```
Comment (dhimmel) [inline hetio/neo4j.py]: Let's open a stack overflow issue with the `neo4j` tag on whether we can do a sum aggregation and then combine that with the pre-aggregated result table all in Cypher.

File: hetio/neo4j.py [hunk 7]
```diff
@@ -281,6 +281,171 @@ def construct_dwpc_query(metarels, property='name', join_hint='midpoint', index_
...
+        if join_hint is True or join_hint == 'midpoint':
+            join_hint = len(metarels) // 2
+        join_hint = int(join_hint)
+        assert join_hint >= 0
+        assert join_hint <= len(metarels)
+        using_query += "\nUSING JOIN ON n{}".format(join_hint)
+
+    # Unique node constraint (pevent paths with duplicate nodes)
+    if unique_nodes == 'nested':
+        unique_nodes_query = '\nAND ALL (x IN nodes(path) WHERE size(filter(z IN nodes(path) WHERE z = x)) = 1)'
```
Comment (dhimmel) [inline hetio/neo4j.py]: Is it possible we can put all the code duplicated by `construct_dwpc_query` in local functions that both methods call?

File: test/neo4j_test.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    'w': damping_exponent,
+    }
+    with driver.session() as session:
+        results = session.run(query, params)
+        results = results.single()
+        assert results
+
+    dwpc = results['DWPC']
+
+    assert dwpc == pytest.approx(0.03287590886921623)
```
Comment (dhimmel) [inline test/neo4j_test.py]: Let's also add tests that evaluate that the expected Cypher query is exported for a given metagraph, but don't actually execute the Cypher.

**Comment by ben-heil** [comment]
Addressed the comments and added the code to #32. I would have created a new pull request, but Github automatically updated the existing one when I pushed to my personal account
